### PR TITLE
add BPF related kernel config

### PR DESCRIPTION
**Pull request by pwang7**
add BPF related kernel config flags, so that people can run BPF with customized kernel in Kata

### CODE REVIEW

**Comment by devimc** [comment]
thanks @pwang7 - Could you please elaborate on your commit message? why we need this and what are you fixing?
take a look to this https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#examples

I  wonder what could be the performance impact of this change

/test-qemu-metrics
/test-metrics

**Comment by pwang7** [comment]
> what

@devimc  Thanks for your reply!

This PR is to enable BPF related features in kernel, when building a customized kernel for Kata.

Without these BPF related features enabled in kernel, many BPF programs cannot run in Kata, such as using BPF for tracing, network filtering, etc.

The main purpose of this PR is to facilitate BPF developers to run their BPF code in Kata. Since BPF is of active development in kernel community, and many BPF related new features require very recent kernel 5.8+. With this PR to enable BPF features in kernel, it's easy to run and test BPF programs in Kata with multiple customized new kernel versions. Otherwise BPF developers have to install multiple kernel versions to their dev machines and switch to different kernels by reboot.

**Comment by devimc** [comment]
Thanks @pwang7 

> The main purpose of this PR is to facilitate BPF developers to run their BPF code in Kata. Since BPF is of active development in kernel community, and many BPF related new features require very recent kernel 5.8+. With this PR to enable BPF features in kernel, it's easy to run and test BPF programs in Kata with multiple customized new kernel versions. Otherwise BPF developers have to install multiple kernel versions to their dev machines and switch to different kernels by reboot.

That makes sense, but since not all users/developers may need BPF, it should be optional.
would you mind reworking your pull request to make BPF optional? It would be great if you can add an option (--bpf) to `build-kernel.sh` to enable it.
Take a look to this PR https://github.com/kata-containers/packaging/pull/938/files -  it can help you as reference.

Thanks

**Comment by pwang7** [comment]
> Thanks @pwang7
> 
> > The main purpose of this PR is to facilitate BPF developers to run their BPF code in Kata. Since BPF is of active development in kernel community, and many BPF related new features require very recent kernel 5.8+. With this PR to enable BPF features in kernel, it's easy to run and test BPF programs in Kata with multiple customized new kernel versions. Otherwise BPF developers have to install multiple kernel versions to their dev machines and switch to different kernels by reboot.
> 
> That makes sense, but since not all users/developers may need BPF, it should be optional.
> would you mind reworking your pull request to make BPF optional? It would be great if you can add an option (--bpf) to `build-kernel.sh` to enable it.
> Take a look to this PR https://github.com/kata-containers/packaging/pull/938/files - it can help you as reference.
> 
> Thanks

Just updated according to your suggestion, please review.

File: kernel/README.md [hunk 1]
```diff
@@ -44,6 +44,7 @@ Commands:
 Options:
 
 	-c <path>   : Path to config file to build a the kernel.
+	-b          : Enable BPF related features in kernel.
```
Comment (jodh-intel) [inline kernel/README.md:47]: Nit: This hasn't been added to the list of options alphabetically - it should come first.

File: kernel/build-kernel.sh [hunk 2]
```diff
@@ -90,6 +92,7 @@ Commands:
 Options:
 
 	-c <path>   : Path to config file to build a the kernel.
+	-b          : Enable BPF related features in kernel.
```
Comment (jodh-intel) [inline kernel/build-kernel.sh:95]: Same comment as above. Not necessary for this PR, but I'd personally remove the hard-coded output of the script in the README and tell users they can run `build-kernel.sh -h` for further info in the doc - easier maintenance! :smile:

File: kernel/build-kernel.sh [hunk 3]
```diff
@@ -188,7 +191,10 @@ get_kernel_frag_path() {
 
 	local arch_configs="$(ls ${arch_path}/*.conf)"
 	# Exclude configs if they have !$arch tag in the header
-	local common_configs="$(grep "\!${arch}" ${common_path}/*.conf -L)"
+	local common_configs="$(grep "\!${arch}" ${common_path}/*.conf -L | sed '/bpf.conf/d')"
+	if [[ ${enable_bpf_kernel_configs} == "true" ]]; then
```
Comment (jodh-intel) [inline kernel/build-kernel.sh:195]: Nit: You don't need to use double square brackets here.

File: kernel/build-kernel.sh [hunk 4]
```diff
@@ -188,7 +191,10 @@ get_kernel_frag_path() {
...
+		common_configs="$(grep "\!${arch}" ${common_path}/*.conf -L)"
+	fi
```
Comment (jodh-intel) [inline kernel/build-kernel.sh:197]: Nit: Missing blank line after `fi`.

File: kernel/configs/fragments/common/bpf.conf [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# for tc filters
```
Comment (jodh-intel) [inline kernel/configs/fragments/common/bpf.conf:1]: How about adding a 1 line comment explaining what the file is for (using the term Berkeley Packet Filter than just bpf)?

**Review by jodh-intel — APPROVED** [review]
Thanks @pwang7.

**Comment by pwang7** [comment]
> Thanks @pwang7.

Just updated according to your review comments.

**Review by devimc — CHANGES_REQUESTED** [review]
thanks @pwang7 - could you please update your commit message?, it needs a `fixes #NNNN` and a body, see https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#examples

**Comment by pwang7** [comment]
> thanks @pwang7 - could you please update your commit message?, it needs a `fixes #NNNN` and a body, see https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#examples

Do I need to file an issue first? Since I don't have issue number now.

**Comment by devimc** [comment]
> Do I need to file an issue first? Since I don't have issue number now.

yes, please

**Comment by pwang7** [comment]
> > Do I need to file an issue first? Since I don't have issue number now.
> 
> yes, please

Updated.

**Review by devimc — APPROVED** [review]
thanks @pwang7

**Pull request merged**